### ISSUE TITLE
Index file nodes for rust crates [phab:D941]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1263]
_Created by **djrenren** at 2016-06-22 23:14:13_: 

A barebones implementation of a lint-plugin-based indexer for rust.

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-06-22 23:30:48_:

  - Remove wildcard from Cargo dependencies

**Comment by kythe** [comment]
_**zarko** wrote at 2016-06-22 23:45:12_:

How should I build and run the tests? `bazel test //...` didn't run anything Rust-related, then:

```
bazel build //kythe/rust/indexer:libindexer
INFO: Found 1 target...
ERROR: /usr/local/google/home/zarko/kythe-trunk/kythe/rust/indexer/BUILD:5:1: Executing genrule //kythe/rust/indexer:_libindexer-k8 failed: bash failed: error executing command /bin/bash -c ... (remaining 1 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
tools/build_rules/rust/run_cargo.sh: line 6: cd: kythe/rust/indexer/target/debug/: No such file or directory
Target //kythe/rust/indexer:libindexer failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.534s, Critical Path: 0.04s
```

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-06-22 23:49:03_:

zarko: Tests can currently be run as: 
```
bazel test //kythe/rust/indexer/testdata:basic
```

As to your issue I can't reproduce here it looks like cargo build is failing silently. Could you comment the result of using `--verbose_failures`

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-06-22 23:57:31_:

  - Make run_cargo.sh cd properly

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-06-23 00:26:32_:

  - Actually run the tests...

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

This comment's cut off.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

Would it be better to define constants in a separate module instead of using inline string literals? See e.g. https://github.com/google/kythe/blob/master/kythe/go/util/schema/schema.go

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

nit: add a period after complete sentences (here and elsewhere).

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

nit: alphabetize imports?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

nit: avoid unnecessary linebreaks.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

Generally we've found it useful in indexers to limp forward rather than crash outright when something weird happens. Maybe you'll be able to output data for other files?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

This file needs a copyright header.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

Is this necessary (if you're defaulting to an empty corpus in your indexer)?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

I don't think `CLANG_STANDARD` is relevant here.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

This also needs updated.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

You might want to start out with this set to `False`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

This is an interesting approach; how do you plan on handling links across corpora?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

base64 encoding is a requirement of the JSON wire format, not of the schema in general. Is this a layering violation?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

Edge kinds are strings (and must be utf8-encoded).

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-23 00:59:52_:

Do you need to keep these around in memory? How do you feel about using a streaming output interface that eagerly writes `Entry`s? In that case you might not have to do quite so much copying.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-06-23 00:59:52_:

Can the build rules be instructed to fail or print a warning if a nightly Rust isn't installed?

Regarding style mechanics: it seems like the Rust community standard is 99 columns and 4-space tabs. I don't see why we shouldn't stick to this (+ mandatory rust-fmt).

It would be useful to include a README.md that describes how the indexer should be invoked (and its various command-line and environment options).

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

haha whoops was supposed to be a continuation of the same sentence. Lost track of my grammar halfway through

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

Hadn't thought about that really... I suppose regardless of how I implement it, I'll need to define some way of noting and detecting corpora boundaries. I suppose I could encode that info into a CorporaConfig™ that encodes the base corpus name and then the required information to detect boundaries (presumably at crate and ffi boundaries?)

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

These actually are effectively streaming.  (see `src/pass.rs`). The interface of this function allows it to be separated from the driving logic. I'll have to look into the performance concerns here.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

yes

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

Whoops, that's what I get for trying to commit code I'm not currently running. Yeah all strings in rust are UTF-8

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

The fact_value is a private member of the struct so once the object is created it's inaccessible. I'll have to check rustc, I may be able to store it as bytes on the node if rustc will base64 encode on the way out.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

fair :(

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

so for crates and use statements I can totally do that. Modules not so much because order matters.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

Sure thing. Just figured an inability to read the file was indicative of a big enough problem to abort. I'll bring it in line with everything else.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

Huh didn't see it in our other build files, care to double check that?  The CXX indexer and indexer testdata BUILD files don't include one.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

Because the tests are run locally, I believe it's possible an environment variable could leak in.

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-23 01:20:49_:

Copy pasta got me again.

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-06-23 01:20:49_:

Will post an update tomorrow with the listed fixes.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-06-23 20:28:24_:

why? does order matter?

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-24 18:46:07_:

Followed the example of cxx_indexer_test where the first element is the entrypoint for the compiler.

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-06-24 18:48:39_:

  - Address feedback. move constants into their own modules.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-26 18:10:32_:

That sounds reasonable (for a future diff). I also think that it's fine to assign corpora to crates (and not some finer unit of code).

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-06-26 18:10:32_:

If you're only ever going to dump `Entry`s to output, wouldn't it make more sense to pass in an interface (possibly with node, edge, and ordinal-edge functions) that could do so without building intermediate data structures?

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-06-27 17:21:49_:

Yeah definitely could. Was avoiding this strategy at first because I wanted to minimize things I'd have to mock in tests. but yeah that'd probably be a solid performance gain. I can do that.  (Plus most of the testing will be done with the verifier anyway)

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-06-27 20:13:39_

**Comment by kythe** [comment]
_**djrenren** closed via commit d1aca5ad0603d723aab8f7913e3f6cca378954d0 at 2016-06-27 20:39:01_
